modules/nf-cid/src/main/nextflow/data/cid/CidHistoryFile.groovy

```diff
@@ -0,0 +1,170 @@
+import java.nio.channels.FileLock
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardOpenOption
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import nextflow.extension.FilesEx
+/**
+ * File to store a history of the workflow executions and their corresponding CIDs
+ *
+ * @author Jorge Ejarque <jorge.ejarque@seqera.io>
+ */
+@Slf4j
+@CompileStatic
+class CidHistoryFile implements CidHistoryLog {
+
+    Path path
+
+    CidHistoryFile(Path file) {
+        this.path = file
+    }
+
+    void write(String name, UUID key, String runCid, Date date = null) {
+        assert key
+
+        withFileLock {
+            def timestamp = date ?: new Date()
+            log.trace("Writting record for $key in CID history file ${FilesEx.toUriString(this.path)}")
+            path << new CidHistoryRecord(timestamp, name, key, runCid).toString() << '\n'
+        }
+    }
+
+    void updateRunCid(UUID sessionId, String runCid) {
+        assert sessionId
+
+        try {
+            withFileLock { updateRunCid0(sessionId, runCid) }
+        }
+        catch (Throwable e) {
+            log.warn "Can't update CID history file: ${FilesEx.toUriString(this.path)}", e.message
+        }
+    }
+
+    List<CidHistoryRecord> getRecords(){
+        List<CidHistoryRecord> list = new LinkedList<CidHistoryRecord>()
+        try {
+            withFileLock { this.path.eachLine {list.add(CidHistoryRecord.parse(it)) } }
+        }
+        catch (Throwable e) {
+            log.warn "Can't read records from CID history file: ${FilesEx.toUriString(this.path)}", e.message
+        }
+        return list
+    }
+
+
+    CidHistoryRecord getRecord(UUID id) {
+        assert id
+
+        for (String line : this.path.readLines()) {
+            def current = line ? CidHistoryRecord.parse(line) : null
+            if (current.sessionId == id) {
+                return current
+            }
+        }
+        log.warn("Can't find session $id in CID history file ${FilesEx.toUriString(this.path)}")
+        return null
+    }
+
+
+    private void updateRunCid0(UUID id, String runCid) {
+        assert id
+        def newHistory = new StringBuilder()
+
+        for( String line : this.path.readLines()) {
+            try {
+                def current = line ? CidHistoryRecord.parse(line) : null
+                if (current.sessionId == id) {
+                    log.trace("Updating record for $id in CID history file ${FilesEx.toUriString(this.path)}")
+                    final newRecord = new CidHistoryRecord(current.timestamp, current.runName, current.sessionId, runCid)
+                    newHistory << newRecord.toString() << '\n'
+                } else {
+                    newHistory << line << '\n'
+                }
+            }
+            catch (IllegalArgumentException e) {
+                log.warn("Can't read CID history file: ${FilesEx.toUriString(this.path)}", e.message)
```

Comment:
This should be review to make it work without the need for locks.

Reply:
I followed the same as the executions log file. Two executions could write the log. The other option is converting the `.history ` to a folder and have a file per run, the history log will be the list of the content of the files.

Reply:
> The other option is converting the .history  to a folder and have a file per run, the history log will be the list of the content of the files

I was thinking the same